### PR TITLE
[precompile] Port the rest of test_precompile.py to the on-disk capture()/load() pair

### PR DESCRIPTION
**Pull request by bobrenjc93**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.17.0) (oldest at bottom):
* #197353
* #197352
* #197347
* #197345
* #197344
* #197456
* #197763
* #197767
* #197766
* #197343
* #197351
* #197839
* #197998
* #197849
* #197844
* #197689
* #197720
* #197600
* __->__ #197388
* #197387
* #197455
* #197341
* #197342
* #197348
* #197349

**Summary.** The commit below ported the first half of `TestPrecompile` to `torch._precompile.capture()` and `load()` through the `_CaptureToFiles` and `_load_pair` adapters; this commit ports the remaining methods the same way. Tests that pinned the shape of the in-memory `(python_code, cache)` return now read that pair back from the artifact and cache files, and tests that loaded it call `_load_pair`, which stages the two halves to disk. The few callable-API tests with no file-pair equivalent, such as the one asserting `load` still reads an in-memory pair, are kept as they are for now and retire with the callable in the module commit above, so no test is dropped -- but this is not a pure re-spelling of assertions: the adapters flip the grad mode the non-training captures trace under, and capture serving the step for real needs eight `zero_grad(set_to_none=True)` resets, both described below. Tests that pin the exact cache envelope key set or the public qualname keep their current expectations; the module commit updates them when it adds the `tracer` tag. No library code changes in this commit.

## The port

```python
python_code, cache = torch.compiler.precompile(fn, x, backend="eager")
self.assertEqual(torch.compiler.precompile.load(python_code, cache)(x), fn(x))

with _CaptureToFiles(fn, backend="eager") as cap:
    cap(x)
python_code, cache = cap.result()
self.assertEqual(_load_pair(python_code, cache)(x), fn(x))
```

## Grad mode flips for the non-training conversions

The callable API wrapped the trace in `torch.enable_grad()` unconditionally; `capture()` lets the caller pick, and `_CaptureToFiles` defaults `training=False`:

```python
with torch.enable_grad():
    compiled._compile(example_inputs)

with torch.enable_grad() if self._training else torch.no_grad():
    self._module._compile(args)
```

So the 45 conversions here that do not pass `training=True` (56 `_CaptureToFiles(...)` sites, 11 of them training) now trace under `no_grad` where they used to trace under `enable_grad`. AOTAutograd dispatch is unaffected: `compile_to_python` pins `no_grad` internally ("no_grad pins the inference path"), so both spellings always took the inference path. The `make_fx` trace itself is not: a grad-enabled trace picks up autograd bookkeeping (e.g. a dead `aten.alias`), so the eager-backend artifact text these tests render is no longer byte-identical to the pre-port text -- which is why `test_callable_api_load_still_reads_an_in_memory_pair` has to pass `training=True` to keep byte-matching the callable API's `python_code`. No assertion in this diff is sensitive to the difference: no ported site pins the graph body or the whole artifact text; the only text assertions are `assertIn` checks on metadata lines (`PARAM_NAMES`, `BUFFER_NAMES`, `IN_SPEC`, `USER_INPUT_SHAPES`/`USER_INPUT_BOUNDS`, `MODULE_POSITIONS`), which the grad mode does not change.

## Capture serves the step for real, so training sites reset .grad

Capture runs the captured step once for real and serves the artifact (the caller-driven contract), so a `training=True` capture leaves one backward's worth of gradient on the example model. Seven capture sites therefore reset it (eight calls -- one site drives two models), each placed before the reference gradients are taken or the loaded artifact is served, so the reference and the artifact each measure exactly one backward:

```python
with _CaptureToFiles(lambda model, t: model(t).sum().backward(), training=True) as cap:
    cap(m, t)
code, cache = cap.result()
m.zero_grad(set_to_none=True)  # capture already ran one backward
ref = copy.deepcopy(m)
ref(t).sum().backward()
_load_pair(code, cache)(m, t)
```

The same effect is now spelled out in `test_grad_identity_preserved_across_precompile`, where the served scatter's in-place `p.grad += g` is a second reason its `assertIs` holds.

## Where an assertion changes

`test_plain_function`'s `assertIsInstance(code, str)` / `assertIsInstance(cache, bytes)` would become tautologies under the port, since `cap.result()` reads a `str` off the artifact file and `bytes` off the cache file by construction. They are replaced by assertions the file pair can violate: both files exist and are non-empty once the block exits, and the cache parses under the safe unpickler with a compiled artifact in it -- a `None` artifact would silently push the loads below onto the inlined path. The envelope's key set is device-independent and stays pinned by the envelope tests in `TestPrecompile`, so this device-parametrized test does not re-assert it. `test_batchnorm_train_buffer_mutation` retargets three assertions: it pinned the example module's `running_mean` / `running_var` / `num_batches_tracked` against pre-capture clones, which held while capture only traced; under the adapter the capture call also serves the artifact once on that model, so the port compares the stats with the reference model's single eager call instead. That pins the same invariant at least as tightly: a real-tensor trace would advance the stats a second time, and `num_batches_tracked` is an exact integer, so the pin cannot pass by tolerance. The shared `_default_and_inlined_loaders` helper switches from the callable's in-memory `load` to `_load_pair` here, together with its three callers (`test_param_shape_mismatch_rejected`, `test_param_dtype_mismatch_rejected`, `test_buffer_shape_dtype_mismatch_rejected`), which this commit ports. `test_param_layout_specialization_rejected_inductor`'s local `loaders()` generator becomes a call to the shared `_default_and_inlined_loaders` helper it had grown identical to, moving its subtest label from `cached` to `default`.

## Test Plan

```
python test/test_precompile.py
python -m pyflakes test/test_precompile.py
lintrunner test/test_precompile.py
```

All tests pass except the CUDA and triton-only ones, which error in this CPU-only environment on the parent commit too.

Authored with an AI assistant.

### CODE REVIEW

**Comment by bobrenjc93** [comment]
@claude

**Comment by bobrenjc93** [comment]
Closing: the author is taking a different direction with this work. Nothing here landed; the branch is left in place.
